aten/src/ATen/native/cuda/Reduce.cuh

```diff
@@ -508,8 +538,38 @@ static void launch_reduce_kernel(const ReduceConfig& config, const R& reduction)
   AT_CUDA_CHECK(cudaGetLastError());
 }
 
+struct AccumulationBuffer {
+  AccumulationBuffer() {}
+
+  AccumulationBuffer(size_t acc_stride, size_t out_stride, char* out_ptr, int64_t size) {
+    out_ptr_ = (char*)out_ptr;
+    if (out_stride > acc_stride) {
+      acc_ptr_ = (char*)out_ptr;
+      size_factor_ = 1;
+    } else {
+      auto& allocator = *at::globalContext().getTHCState()->cudaDeviceAllocator;
+      buffer_ = allocator.allocate(size);
+      acc_ptr_ = (char*)buffer_.get();
+      size_factor_ = float(out_stride) / acc_stride;
+    }
+  }
+
+  char* get_acc_slice(char* out_ptr) {
+    if (size_factor_ == -1) {
+      return nullptr;
+    }
+    return acc_ptr_ + (int)((out_ptr - out_ptr_) * size_factor_);
```

Comment:
I'm nervous here and in other places about pointer arithmetic involving floats. Also, I'm nervous about casting to (int) because it might conceivably overflow for large outputs and size_factors_.

Reply:
Fixed pointer arithmetic with casting and floats to properly use `size_t` instead.